src/fpm_sources.f90

```diff
@@ -65,13 +65,16 @@ subroutine add_sources_from_dir(sources,directory,scope,with_executables,recurse
 
     integer :: i
     logical, allocatable :: is_source(:), exclude_source(:)
+    logical :: recurse_
     type(string_t), allocatable :: file_names(:)
     type(string_t), allocatable :: src_file_names(:)
     type(string_t), allocatable :: existing_src_files(:)
     type(srcfile_t), allocatable :: dir_sources(:)
 
+    recurse_ = .true.
+    if (present(recurse)) recurse_ = recurse
     ! Scan directory for sources
-    call list_files(directory, file_names,recurse=merge(recurse,.true.,present(recurse)))
+    call list_files(directory, file_names,recurse=recurse_)
```

Comment:
The `recurse` argument for `list_files` is also optional, so `recurse` can be passed straight to `list_files` without extra logic. (The original logic with `merge` was unnecessary.)

> [A non-present optional dummy argument] may be supplied as an actual argument corresponding to an optional dummy argument, which is then considered not to be present.

_p4, sec 15.5.2.12 18-007r1 (standard interpretation doc)_

Reply:
But this would change the behavior of this subroutine as `list_files` defaults to false if *recurse* is not present.

https://github.com/fortran-lang/fpm/blob/bb449174631be989fae08532de31e94bac5c5ac3/src/fpm_filesystem.F90#L472-L473

Reply:
Ah yes, so you are right. That's annoyingly inconsistent, but in that case your change is otherwise fine.